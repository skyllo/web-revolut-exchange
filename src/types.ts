export enum Pocket {
  EUR = 'EUR',
  GBP = 'GBP',
  USD = 'USD',
}

export interface PocketData {
  name: Pocket;
  balance: number;
  symbol: string;
  title: string;
  countryCode: string;
}

export type InputDirection = 'from' | 'to';

export interface Rate {
  rate: number;
  fromPocket: Pocket;
  toPocket: Pocket;
}
