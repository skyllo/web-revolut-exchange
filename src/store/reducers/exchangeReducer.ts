/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Pocket } from '../../types';

export interface ExhangeState {
  fromPocket: Pocket;
  toPocket: Pocket;
}

const initialState: ExhangeState = {
  fromPocket: Pocket.GBP,
  toPocket: Pocket.EUR,
};

function switchPockets(state: ExhangeState) {
  const { fromPocket, toPocket } = state;
  state.fromPocket = toPocket;
  state.toPocket = fromPocket;
}

export const exchangeSlice = createSlice({
  name: 'exchange',
  initialState,
  reducers: {
    swapPockets(state) {
      switchPockets(state);
    },
    setFromPocket(state, action: PayloadAction<Pocket>) {
      const { fromPocket, toPocket } = state;
      const { payload: pocket } = action;
      // swap or switch to a new pocket when selected
      if (fromPocket === pocket || toPocket === pocket) {
        switchPockets(state);
      } else {
        state.fromPocket = action.payload;
      }
    },
    setToPocket(state, action: PayloadAction<Pocket>) {
      const { fromPocket, toPocket } = state;
      const { payload: pocket } = action;
      // swap or switch to a new pocket when selected
      if (fromPocket === pocket || toPocket === pocket) {
        switchPockets(state);
      } else {
        state.toPocket = action.payload;
      }
    },
  },
});

export const { setFromPocket, setToPocket, swapPockets } = exchangeSlice.actions;

export default exchangeSlice.reducer;
