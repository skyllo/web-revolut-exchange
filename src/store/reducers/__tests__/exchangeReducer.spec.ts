import { Pocket } from '../../../types';
import exchangeReducer, { setFromPocket, setToPocket, swapPockets } from '../exchangeReducer';

describe('exchangeReducer', () => {
  it('should return the initial state', () => {
    expect(exchangeReducer(undefined, { type: '' })).toMatchSnapshot();
  });

  it('should update fromPocket if different', () => {
    expect(
      exchangeReducer(
        {
          fromPocket: Pocket.GBP,
          toPocket: Pocket.EUR,
        },
        setFromPocket(Pocket.EUR)
      )
    ).toMatchSnapshot();
  });

  it('should switch fromPocket if same', () => {
    expect(
      exchangeReducer(
        {
          fromPocket: Pocket.GBP,
          toPocket: Pocket.EUR,
        },
        setFromPocket(Pocket.GBP)
      )
    ).toMatchSnapshot();
  });

  it('should update toPocket if different', () => {
    expect(
      exchangeReducer(
        {
          fromPocket: Pocket.USD,
          toPocket: Pocket.EUR,
        },
        setToPocket(Pocket.GBP)
      )
    ).toMatchSnapshot();
  });

  it('should switch toPocket if same', () => {
    expect(
      exchangeReducer(
        {
          fromPocket: Pocket.USD,
          toPocket: Pocket.EUR,
        },
        setToPocket(Pocket.EUR)
      )
    ).toMatchSnapshot();
  });

  it('should swap pockets', () => {
    expect(
      exchangeReducer(
        {
          fromPocket: Pocket.EUR,
          toPocket: Pocket.USD,
        },
        swapPockets()
      )
    ).toMatchSnapshot();
  });
});
