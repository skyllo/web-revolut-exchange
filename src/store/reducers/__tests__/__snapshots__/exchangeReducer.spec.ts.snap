// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exchangeReducer should return the initial state 1`] = `
Object {
  "fromPocket": "GBP",
  "toPocket": "EUR",
}
`;

exports[`exchangeReducer should swap pockets 1`] = `
Object {
  "fromPocket": "USD",
  "toPocket": "EUR",
}
`;

exports[`exchangeReducer should switch fromPocket if same 1`] = `
Object {
  "fromPocket": "EUR",
  "toPocket": "GBP",
}
`;

exports[`exchangeReducer should switch toPocket if same 1`] = `
Object {
  "fromPocket": "EUR",
  "toPocket": "USD",
}
`;

exports[`exchangeReducer should update fromPocket if different 1`] = `
Object {
  "fromPocket": "EUR",
  "toPocket": "GBP",
}
`;

exports[`exchangeReducer should update toPocket if different 1`] = `
Object {
  "fromPocket": "USD",
  "toPocket": "GBP",
}
`;
